test/dist_autograd_test.py

```diff
@@ -220,6 +223,28 @@ def test_rpc_complex_args(self):
             dst_rank = (self.rank + 1) % self.world_size
             self.assertEqual(worker_ids[0], dst_rank)
 
+
+    @dist_init
+    def test_context_cleanup_many_workers(self):
+        global known_context_ids
+        dst_ranks = {rank for rank in range(self.world_size) if rank != self.rank}
+        with dist_autograd.context() as context_id:
+            t1 = torch.ones(3, 3, requires_grad=True)
+            t2 = torch.zeros(3, 3, requires_grad=True)
+            for dst_rank in dst_ranks:
+                ret = rpc.rpc_sync("worker{}".format(dst_rank), torch.add, args=(t1, t2))
+                rpc.rpc_sync("worker{}".format(dst_rank), store_context_id, args=(context_id,))
+        # now let each worker finish with their cleanup RPCs
+        rpc.sync_rpc()
```

Comment:
dist.barrier().

Reply:
See above conversation regarding this

Reply:
One way to fix this would be to have loop with a timeout which checks for the context and sleeps for a while if the context still exists. If we exit this loop and the timeout has expired, we fail the test. The timeout could be 10s to be conservative. An example:

```
start = time.time()
bool success = False
while time.time() - start < 10:
  if not dist_autograd.has_context_id(context_id):
    success = True
    break
self.assertTrue(success)
```

This structure can also be moved into a helper function that we can reuse in other tests.